storage/remote/client.go

```diff
@@ -38,30 +39,71 @@ const maxErrMsgLen = 256
 
 var userAgent = fmt.Sprintf("Prometheus/%s", version.Version)
 
-// Client allows reading and writing from/to a remote HTTP endpoint.
-type Client struct {
+// client allows reading and writing from/to a remote HTTP endpoint.
+type client struct {
 	remoteName string // Used to differentiate clients in metrics.
 	url        *config_util.URL
 	client     *http.Client
 	timeout    time.Duration
+
+	readQueries prometheus.Gauge
 }
 
-// ClientConfig configures a Client.
+// ClientConfig configures a client.
 type ClientConfig struct {
 	URL              *config_util.URL
 	Timeout          model.Duration
 	HTTPClientConfig config_util.HTTPClientConfig
 }
 
-// NewClient creates a new Client.
-func NewClient(remoteName string, conf *ClientConfig) (*Client, error) {
+// TODO(bwplotka): Add streamed chunked remote read (add issue link).
+type ReadClient interface {
+	Read(ctx context.Context, query *prompb.Query) (*prompb.QueryResult, error)
+}
+
+// NewReadClient creates a new client for remote read.
+// Caller has to ensure there are no duplicated clients.
+func NewReadClient(reg prometheus.Registerer, name string, conf *ClientConfig) (ReadClient, error) {
 	httpClient, err := config_util.NewClientFromConfig(conf.HTTPClientConfig, "remote_storage", false)
 	if err != nil {
 		return nil, err
 	}
 
-	return &Client{
-		remoteName: remoteName,
+	readQueries := prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "remote_read_queries",
+			Help:      "The number of in-flight remote read queries.",
+			ConstLabels: prometheus.Labels{
+				remoteName: name,
+				endpoint:   conf.URL.String(),
+			},
+		},
+	)
+	if reg != nil {
+		reg.Unregister(readQueries)
```

Comment:
This doesn't seem right, you should never have to unreigster outside of tests.

Reply:
Hm, what're rationales behind this?

Because otherwise I would need to move this gauge to storage (which is fair I guess)

Reply:
A given metric should exist for the lifetime of a process, if you're unregistering then that implies something is off with the metric lifecycle.

Reply:
oke will fix